Doip

wosym opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

kaspar030 committed:
pkg: micropython: initial import

kaspar030 committed:
examples/micropython: initial micropython example application

aabadie committed:
sys/progress_bar: add module for configurable progress bars

aabadie committed:
tests/progress_bar: add test application

benpicco committed:
lpc2k_pgm: add LPC2388

The device is very simmilar to the LPC2387

Commit:
pseudomodules: add sock_dtls

Commit:
pkg/tinydtls: implement DTLS sock API

Commit:
examples: add DTLS sock example application

ant9000 committed:
USBUS CDC ECM: host and MCU mac addresses should differ

benpicco committed:
tests/periph_pm: use button as a wake-up source

To test if GPIO interrupts can wake the CPU from sleep, configure
BTN0 (if availiable) as a wake-up source.

Pressing the buttong should wake up the CPU.

wosym committed:
Implemented STM32 Ethernet driver for nucleo-f207zg

wosym committed:
Added the udp-client/server, heavily based on the RIOT tutorial. This one already uses the socket_abstraction (except for 1 conditional), so this is probably a good point to start the port to LWIP

wosym committed:
Added the udp-client/server, heavily based on the RIOT tutorial. This one already uses the socket_abstraction (except for 1 conditional), so this is probably a good point to start the port to LWIP

wosym committed:
Basic send packet over IP with LWIP on IPv6

wosym committed:
Basic send packet over IP with LWIP on IPv6

wosym committed:
Adapt LWIP bootstrap sequence to IPv4 functions

wosym committed:
Adapt LWIP bootstrap sequence to IPv4 functions

wosym committed:
improve IPV4 lwip gluecode by making the IP adresses selectable in the Makefile

Auto choosing adresses did not work.
Also finished up the ifconfig for ipv4

wosym committed:
improve IPV4 lwip gluecode by making the IP adresses selectable in the Makefile

Auto choosing adresses did not work.
Also finished up the ifconfig for ipv4

wosym committed:
Fix ip_send for ipv4

wosym committed:
Fix ip_send for ipv4

aabadie committed:
tests/stdin: refactor test application

The test application now prints in a loop the input character. In case
stdin is not ready yet after startup this lets the possibility to try to
send several time a character before failing.
The automatic test is now more robust on platforms where stdin takes
time before it gets in a ready state (some AVR, hifive).

wosym committed:
Tried to make a DoIP module. Still needs a lot of work

wosym committed:
Tried to make a DoIP module. Still needs a lot of work

wosym committed:
Implement first DoIP API functions and test them

Still needs cleaning up, and we'll probably rename the functions later
on to use all lowercase.

wosym committed:
Implement first DoIP API functions and test them

Still needs cleaning up, and we'll probably rename the functions later
on to use all lowercase.

wosym committed:
Change naming convention to lowercase doip (in contrary to the spec)

wosym committed:
Change naming convention to lowercase doip (in contrary to the spec)

wosym committed:
change function name (abstraction)

wosym committed:
change function name (abstraction)

wosym committed:
fix small bug with payload length

wosym committed:
fix small bug with payload length

wosym committed:
change ip's to link local

wosym committed:
change ip's to link local

wosym committed:
Add receive after UDP transmit

wosym committed:
Add receive after UDP transmit

wosym committed:
Add print_parsed function + cleanup code

wosym committed:
Add print_parsed function + cleanup code

wosym committed:
try to add DHCP to lwip

wosym committed:
try to add DHCP to lwip

smlng committed:
pkg/jerryscript: set -Wno-conversion for FreeBSD

Fixes compile issue on FreeBSD because of float to double conversion
by disabling the corresponding compiler warning.

wosym committed:
Make the IP a function parameter for doip_send_udp

wosym committed:
Make the IP a function parameter for doip_send_udp

benpicco committed:
lpc2k_pgm: fix build warnings

GCC now warns when snprintf()ing a buffer that, together with the
format string, might be larger than the destination buffer.

To fix this increase the size of the destination buffer so that
the source buffer and the format string will always fit - 32 bytes
are enough.

wosym committed:
fix dhcp by moving procedure to usercode

wosym committed:
fix dhcp by moving procedure to usercode

wosym committed:
refactor some code, add extra doip_send function

wosym committed:
refactor some code, add extra doip_send function

wosym committed:
Add some documentation to sys/doip

wosym committed:
Add some documentation to sys/doip

wosym committed:
Add comments to doip example

wosym committed:
Add comments to doip example

wosym committed:
make minor change to lwip bootstrap so that non-dhcp still works

wosym committed:
make minor change to lwip bootstrap so that non-dhcp still works

wosym committed:
Fix coding style

wosym committed:
Fix coding style

wosym committed:
Fix coding style to RIOT-style

wosym committed:
Fix coding style to RIOT-style

wosym committed:
cleaned up lwip bootstrap

wosym committed:
cleaned up lwip bootstrap

toonst commented on boards/nucleo-f207zg/include/periph_conf.h:
```diff
@@ -41,6 +41,7 @@ static const dma_conf_t dma_config[] = {
     { .stream = 4 },    /* DMA1 Stream 4 - SPI2_TX */
     { .stream = 14 },   /* DMA2 Stream 6 - USART6_TX */
     { .stream = 6 },    /* DMA1 Stream 6 - USART2_TX */
+    { .stream = 8 },    /* DMA2 Stream 0 - ETH_TX    */ // NOTE: Probeersel... komt uit andere periph_config!
```
rebase on master to get latest periph_conf that includes this already

toonst commented on examples/doip/main.c:
```diff
@@ -0,0 +1,88 @@
+    char line_buf[SHELL_DEFAULT_BUFSIZE];
+
+    puts("Starting doip example");
+    xtimer_usleep(2000000);         //necesarry for some reason when sending UDP
```
there's `xtimer_sleep` which you can pass seconds to. Also use a preprocessor define for all magic values.

toonst commented on exp/Makefile:
```diff
@@ -0,0 +1,65 @@
+###
```
is this exp app still needed?

toonst commented on pkg/lwip/contrib/lwip.c:
```diff
@@ -42,11 +42,19 @@
+#ifdef USE_DHCP
+#include <lwip/dhcp.h>
+#endif
+
```
we should let the user decide whether to use DHCP or not.

wosym committed:
added color.h for doip

wosym committed:
added color.h for doip

wosym committed:
Revert "added color.h for doip"

This reverts commit d3e1c09702ced54e5d7526578b91bf2a5f6be404.

wosym committed:
Revert "added color.h for doip"

This reverts commit d3e1c09702ced54e5d7526578b91bf2a5f6be404.

wosym committed:
add missing color.h

wosym committed:
add missing color.h

maribu committed:
Merge pull request #12868 from benpicco/lpc2k-fix

lpc2k_pgm: fix build warnings, add lpc2388

maribu committed:
drivers/ws281x: Fixed issue in doc

haukepetersen committed:
sys/fmt: move _is_digit and _is_upper to public API

haukepetersen committed:
sys/fmt: add fmt_is_number()

haukepetersen committed:
unittests: add tests for fmt_is_x() functions

haukepetersen committed:
test/kw2xrf: use fmt_is_number()

aabadie committed:
cpu/fe310: provide gpio feature at cpu level

aabadie committed:
boards/hifive1*: remove periph_gpio provided features

There are now provided at cpu level

fjmolinas committed:
Merge pull request #12816 from aabadie/pr/tests/stdin_fix_avr

tests/stdin: fix automatic test on slow platforms

haukepetersen committed:
shell/sc_gnrc_netif: use fmt_is_number()

wosym committed:
Move Eth init back to bootstrap, DHCP works on its own

miri64 committed:
Merge pull request #12865 from smlng/pr_freebsd_jerryscript

pkg/jerryscript: set -Wno-conversion for FreeBSD

kaspar030 committed:
Merge pull request #12550 from aabadie/pr/sys/progress_bar

sys/progress_bar: add module for managing a progress bar in stdout

benpicco committed:
Merge pull request #12710 from benpicco/tests/periph_pm-button

tests/periph_pm: use button as a wake-up source

benpicco committed:
Merge pull request #12876 from maribu/doc_fix

drivers/ws281x: Fixed issue in doc

smlng committed:
Merge pull request #11943 from pokgak/sock_dtls_impl

pkg/tinydtls: add DTLS sock API implementation

aabadie committed:
cpu/fe310: add cpp feature

kaspar030 committed:
Merge pull request #12880 from aabadie/pr/cpu/fe310_factorize_periph_gpio

cpu/fe310: factorize periph_gpio* provided features

toonst commented on examples/doip/main.c:
```diff
@@ -0,0 +1,88 @@
+                      ip_addr);
+        doip_send_udp(0, 0, DOIP_VEHICLE_IDENTIFICATION_REQUEST, NULL, 0,
+                      ip_addr);
+        xtimer_usleep(1000000);
```
xtimer_sleep iso xtimer_usleep, use preprocessor defines for magic values.

toonst commented on examples/doip/main.c:
```diff
@@ -0,0 +1,88 @@
+        dlen = 0;
+    }
+    else {
+        source = 0xe801;
```
preprocessor defines for magic values

toonst commented on sys/Makefile:
```diff
@@ -160,6 +160,9 @@ endif
 endif
+ifneq (,$(filter doip,$(USEMODULE)))
+    DIRS += doip
+endif
```
I think doip belongs under `sys/net`instead of `sys`

toonst commented on sys/doip/Makefile:
```diff
@@ -0,0 +1,4 @@
+ifeq (gnu, $(TOOLCHAIN))
+  CFLAGS += -Wno-cast-function-type
```
Please remove this hack

toonst commented on sys/doip/color.h:
```diff
@@ -0,0 +1,73 @@
+//Regular text
```
this belongs in the app, not in sys

toonst commented on sys/doip/doip.c:
```diff
@@ -0,0 +1,193 @@
+    int ret = 0;
+    sock_udp_ep_t remote = SOCK_IPV4_EP_ANY;
+    sock_udp_ep_t local = SOCK_IPV4_EP_ANY;
+    sock_udp_t sock;
```
this sock can be made part of a `sock_doip_t`

toonst commented on sys/include/doip.h:
```diff
@@ -0,0 +1,180 @@
+/* ####################*/
+
+
+/* ####### DoIP Functions ########## */
```
I miss a `sock_doip_t`here, which holds the underlying udp socket, tcp socket, static buffers and other state. This socket can then be created by a `sock_doip_create` function and closed by a `sock_doip_close`function, similar to the other protocols which implement a socket interface. See eg. `sys/include/net/sock/udp.h`

The biggest benefit of this is that the socket contains all the state of a connection, allowing to open multiple connections.

wosym replied:
@toonst Should the code be POSIX compliant? Because if I'm not mistaken, POSIX has _t reserved for their own structs.

toonst replied:
it's fine to use _t here.

toonst commented on tests/doip/main.c:
```diff
@@ -0,0 +1,57 @@
+
+}
+
+int main(void)
```
This test app is definitely needed, so please fix this.

toonst commented on sys/include/doip.h:
```diff
@@ -0,0 +1,180 @@
+ *
+ * Available service primitives:
+ * - request: request a service from the service provider
+ * - inidcation: service provider informs a service user about an internal event of the network layer or the service request of a peer protocol layer entity service user
```
typo: ìnidcation`

toonst commented on sys/doip/doip.c:
```diff
@@ -0,0 +1,193 @@
+    }
+
+
+    //TODO: make a message_create function?
```
yes! ;-)

toonst commented on sys/doip/doip.c:
```diff
@@ -0,0 +1,193 @@
+    int msglen = 0;         //Total length of message to be sent over UDP
+
+    uint8_t buf[50] = { '\0' };
+    if (sock_udp_create(&sock, &local, NULL, 0)) {
```
move this socket creation to a new `sock_doip_create`function

toonst replied:
A doip socket can be used so the underlying udp socket doesn't have to be created and closed every time you want to send something

toonst commented on sys/doip/doip.c:
```diff
@@ -0,0 +1,193 @@
+
+    puts("======================");
+    sock_udp_close(&sock);
+    xtimer_sleep(1);
```
sleep still needed?

toonst commented on sys/doip/doip.c:
```diff
@@ -0,0 +1,193 @@
+    }
+
+    puts("======================");
+    sock_udp_close(&sock);
```
move closing of the socket to a new `socket_doip_close` function

toonst commented on examples/doip/Makefile:
```diff
@@ -0,0 +1,47 @@
+USEMODULE += lwip_tcp lwip_sock_tcp
+USEMODULE += netdev_default
+USEMODULE += xtimer
+USEMODULE += shell
```
Please group USEMODULE's and other stuff separately.

toonst commented on examples/doip/netutil.c:
```diff
@@ -0,0 +1,63 @@
+    stm32_eth_netdev_setup(&stm32eth);
+    int ret = 0;
+
+    if (netif_add(&netif[0], IP4_ADDR_ANY, IP4_ADDR_ANY, IP4_ADDR_ANY,
```
netif add should be part of the lwip bootstrap code.

toonst commented on examples/doip/netutil.c:
```diff
@@ -0,0 +1,63 @@
+        puts("Could not add stm32_eth device\n");
+        return -1;
+    }
+    ret = dhcp_start(&netif[0]);
```
This should be done from the application. The netif can be retrieved with `netif_find()`

benpicco committed:
Merge pull request #12676 from ant9000/pr/usbus_cdc_ecm-mac_host-fix

USBUS CDC ECM: host and MCU mac addresses should differ

toonst commented on examples/doip/netutil.c:
```diff
@@ -0,0 +1,63 @@
+            }
+        }
+#endif
+#ifndef APP_IPV6
```
Can we try to re-use the MODULE_LWIP_IPV6 instead of adding another preprocessor define?

toonst reviewed (changes requested):
Finished review. I skipped the `exp` app because I think it is a leftover.

Commit:
boards/microduino-corerf: doc SPI and I2C pin mapping

bergzand committed:
Merge pull request #2968 from kaspar030/micropython

pkg: micropython: initial commit

kaspar030 committed:
Merge pull request #12878 from haukepetersen/add_fmt_isnumber

sys/fmt: add fmt_is_number()

benpicco committed:
Merge pull request #12883 from fabian18/microduino-corerf-doc-pin-mapping

boards/microduino-corerf: doc SPI and I2C pin mapping

benpicco committed:
tests/periph_backup_ram: disable test_utils_interactive_sync

The `periph_backup_ram` test expects the CPU to enter Deep Sleep, wake up
(causing a CPU reset) increment a counter and go back to sleep.

Introducing a requirement on interaction after reset breaks the test.

miri64 committed:
gnrc_ipv6: fix source check for loopback address

When the destination address is the loopback address (`::1`) in GNRC
the selected network interface typically is `NULL`, as with GNRC no
loopback interface de facto exists. So the assertion when checking if
the source address is valid if `netif != NULL` fails on that check.
This change fixes that issue by checking if the destination address is
the loopback address, before checking the validity of the source
address.

fjmolinas committed:
Merge pull request #12885 from benpicco/tests-fix_periph_backup_ram

tests/periph_backup_ram: disable test_utils_interactive_sync

benpicco committed:
Merge pull request #12886 from miri64/gnrc_ipv6/fix/loopback-src-check

gnrc_ipv6: fix source check for loopback address

wosym committed:
Apply fixes suggested by code review

wosym committed:
fix merge conflicts

wosym committed:
Try to change doip implementation to use struct. Kernel panic at sock_dup_create() for now --> needs fixing

wosym committed:
Not sure what the best approach is to make this architecture-wise...

wosym committed:
sock_doip_t: fix

wosym committed:
doip: main make IP selectable from shell

wosym committed:
Add socket_doip_close() to public interface

wosym committed:
sys/doip: make seperate doip_create_message() function

wosym committed:
sys/doip: remove unecesarry sleep

wosym committed:
add tcp_connect and _send

wosym committed:
continue tcp part: add receiving part, close, extra recv

wosym committed:
doip: make example more userfriendly and streamlined

wosym committed:
doip: finish up tcp

wosym committed:
fix timeout bug

wosym committed:
Fix shell command names

wosym committed:
Don't add +4 to payload_len if there's no TA and SA
